include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,414 @@
+     *  copyIntensityDeviceToHost should be called before */
+    void collectIntensityOnMaster()
+    {
+        reduce(nvidia::functors::Add(), sumfcoskr_master,
+            sumfcoskr->getHostBuffer().getBasePointer(), n_q,
+            mpi::reduceMethods::Reduce());
+        reduce(nvidia::functors::Add(), sumfsinkr_master,
+            sumfsinkr->getHostBuffer().getBasePointer(), n_q,
+            mpi::reduceMethods::Reduce());
+        reduce(nvidia::functors::Add(), &np_master,
+            np->getHostBuffer().getBasePointer(), 1,
+            mpi::reduceMethods::Reduce());
+        reduce(nvidia::functors::Add(), &nmp_master,
+            nmp->getHostBuffer().getBasePointer(), 1,
+            mpi::reduceMethods::Reduce());
+
+        // Calculate intensity on master
+        if (isMaster)
+        {
+            for (unsigned int i = 0; i < n_q; i++)
+                intensity_master[i] =
+                    (sumfcoskr_master[i] * sumfcoskr_master[i] +
+                        sumfsinkr_master[i] * sumfsinkr_master[i]) /
+                    np_master.x();
+
+            std::stringstream o_step;
+            o_step << currentStep;
+            writeIntensity(intensity_master,
+                "saxsOutput/" + filename_prefix + "_" + o_step.str() + ".dat");
+            writeLog(np_master, nmp_master,
+                "saxsOutput/" + filename_prefix + "_" + o_step.str() + ".log");
+        }
+    }
+
+    // perform all operations to get data from GPU to master
+    void collectDataGPUToMaster()
+    {
+        // collect data GPU -> CPU -> Master
+        copyIntensityDeviceToHost();
+        collectIntensityOnMaster();
+    }
+
+    /**
+     * This functions calls the SAXS kernel. It specifies how the
+     * calculation is parallelized.
+     **/
+    void calculateSAXS(uint32_t currentStep)
+    {
+        this->currentStep = currentStep;
+
+        DataConnector &dc = Environment<>::get().DataConnector();
+        auto particles =
+            dc.get<ParticlesType>(ParticlesType::FrameType::getName(), true);
+
+        // calculate the absolute position of the particles
+        const SubGrid<simDim> &subGrid = Environment<simDim>::get().SubGrid();
+        DataSpace<simDim> globalOffset(subGrid.getLocalDomain().offset);
+
+        constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+            pmacc::math::CT::volume<SuperCellSize>::type::value>::value;
+
+        // initialize variables with zero
+        sumfcoskr->getDeviceBuffer().setValue(0.0);
+        sumfsinkr->getDeviceBuffer().setValue(0.0);
+        np->getDeviceBuffer().setValue(0.0);
+        nmp->getDeviceBuffer().setValue(0.0);
+
+        // calculate q_step
+        q_step[0] = (q_max[0] - q_min[0]) / n_qx;
+        q_step[1] = (q_max[1] - q_min[1]) / n_qy;
+        q_step[2] = (q_max[2] - q_min[2]) / n_qz;
+
+        // PIC-like kernel call of the SAXS kernel
+        PMACC_KERNEL(
+            KernelSaxs< numWorkers >{}
+        )(
+            1,
```

Comment:
Suggested change of parallelism (1):
Here compute the number of blocks as `(n_q + numWorkers - 1) / numWorkers`
(has to be done together with changes (2), (3), etc. later in the comments).